tradeoffs.Rmd

```diff
@@ -0,0 +1,172 @@
+`get_doi_handle()` would be the absolute minimum, but it alone cannot distinguish between API failure modes 1 from 3:
+`httr::stop_for_status()` could be thrown either because a DOI does not exist or because of some other HTTP error code.
+The differences between 1 and 3 are important for users and the downstream logic in the biblids package.
+We thus need an additional function at this interface, `is_doi_found()`, which checks explicity for failure mode 1.
+
+We can test most API failure modes of `get_doi_handle()` and `is_doi_found()` thus:
+
+```{r test-minimal-wrapper}
+library(testthat)
+skip_if_offline()
+# outcome 1
+test_that("Non-existent DOIs error out", {
+  expect_error(get_doi_handle("10.1000/I_DO_NOT_EXIST"))
+  expect_false(is_doi_found("10.1000/I_DO_NOT_EXIST"))
+})
+# outcome 2a
+test_that("Existing DOIs are found", {
+  testthat::local_edition(3)
+  # this snapshot will also capture some change in the API output
+  expect_snapshot_output(get_doi_handle("10.1038/nature14236"))
+  expect_true(is_doi_found("10.1038/nature14236"))
+})
+# outcome 2b
+test_that("Non-existing values error out", {
+  expect_error(
+    get_doi_handle(
+      doi = "10.1038/nature14236",
+      query = list(type = "I_DO_NOT_EXIST")
+    )
+  )
+})
+```
+
+One might think that these tests should be mocked, rather than running against the live API.
+However, because DOI.org is stable, performant and requires no authentication, the added complexity is not worth it in this case.
+Moreover, `get_doi_handle()` and `is_doi_found()` perform minimal work and are essentially I/O functions.
+They may thus require *only* (broad) integration testing against a live API.
+
+We do, however, need a mock to test the remaining outcome 3, which cannot readily be retrieved from the live API.
+
+```{r mock-minimal}
+library(httptest)
+test_that("Other http error codes are captured", {
+  expect_error(without_internet(is_doi_found("10.1038/nature14236")))
+  expect_error(without_internet(get_doi_handle("10.1038/nature14236")))
+})
+```
+
+Happily, we can cheat a little, and simply create *some* sure HTTP status error, without actually reproducing API behavior in any detail.
+The simplest HTTP error to create is just by cutting the internect connection.
+Users and the downstream logic in biblids do not care what the HTTP status error is beyond outcome 1, so this error is as good as any.
+
+We can now build additional exported functions on top of these primitives, such as `get_doi_url()` to resolve a DOI to a URL, but we won't need to test these against the live or mocked API, because that interface is already fully covered.
+Of course, if `get_doi_url()` did some very complicated parsing of the API response (it doesn't), we might want to test that behavior, too.
+But in that case, we should let it accept the API response as a parameter and test that as we would any other function free of side effects.
+
+## Use Your Judgement
```

Comment:
I think it'd also be good to add something about a team of developers: it might help, to find the balance, to have others review your code, to exchange ideas etc. Even just to speak with others at conferences etc. (maybe I'm saying that because I watched your interesting useR! talk :slightly_smiling_face: ).
Adding such a sentence about community here would IMHO make it look less unattainable and scholarly?

Reply:
I think that just bloats the chapter.
Asking for help and getting fresh eyes applies to pretty much everything, no?